### PR TITLE
Add datafile isValid check in feature accessor methods

### LINKED ISSUE #107: NoMethodError: undefined method `get_feature_flag_from_key` for nil:NilClass when calling `get_feature_variable_*`
SDK version 2.0.1

On occasions where fetching a valid datafile fails, any call to a `get_feature_variable_*` method will raise the exception `NoMethodError: undefined method 'get_feature_flag_from_key' for nil:NilClass` instead of failing gracefully and logging.

Within `Project#get_feature_variable_for_type` there is no check for if the project is useable via the [`is_valid` attr_reader](https://github.com/optimizely/ruby-sdk/blob/c1ca00d7a73581fd5fbafff692e343490d8448b3/lib/optimizely.rb#L35) before attempting to access the config

https://github.com/optimizely/ruby-sdk/blob/c1ca00d7a73581fd5fbafff692e343490d8448b3/lib/optimizely.rb#L439-L449

All of the other methods seem to return nil or false if the datafile cannot be parsed

https://github.com/optimizely/ruby-sdk/blob/c1ca00d7a73581fd5fbafff692e343490d8448b3/lib/optimizely.rb#L262-L266

### PR DESCRIPTION
**Pull request by oakbani**
Fixes #107

### CODE REVIEW

File: spec/project_spec.rb [hunk 1]
```diff
@@ -1029,6 +1040,17 @@ class InvalidErrorHandler; end
...
     user_attributes = {}
 
+    it 'should return nil when called with invalid project config' do
+      logger = double('logger')
+      allow(logger).to receive(:log)
+      allow(Optimizely::SimpleLogger).to receive(:new) { logger }
+      invalid_project = Optimizely::Project.new('invalid', nil, spy_logger)
+      expect(invalid_project.get_feature_variable_boolean('boolean_single_variable_feature', 'boolean_variable', user_id, user_attributes))
+        .to eq(nil)
+      expect(logger).to have_received(:log).once.with(Logger::ERROR, 'Provided datafile is in an invalid format.')
```
Comment (mikeng13) [inline spec/project_spec.rb]: Is this right? How come we use similar messages twice?

File: spec/project_spec.rb [hunk 2]
```diff
@@ -1055,6 +1077,17 @@ class InvalidErrorHandler; end
...
     user_attributes = {}
 
+    it 'should return nil when called with invalid project config' do
+      logger = double('logger')
+      allow(logger).to receive(:log)
+      allow(Optimizely::SimpleLogger).to receive(:new) { logger }
+      invalid_project = Optimizely::Project.new('invalid', nil, spy_logger)
+      expect(invalid_project.get_feature_variable_double('double_single_variable_feature', 'double_variable', user_id, user_attributes))
+        .to eq(nil)
+      expect(logger).to have_received(:log).once.with(Logger::ERROR, 'Provided datafile is in an invalid format.')
```
Comment (mikeng13) [inline spec/project_spec.rb]: Shouldn't we only be logging the second message?

File: spec/project_spec.rb [hunk 3]
```diff
@@ -1082,6 +1115,17 @@ class InvalidErrorHandler; end
...
 
+    it 'should return nil when called with invalid project config' do
+      logger = double('logger')
+      allow(logger).to receive(:log)
+      allow(Optimizely::SimpleLogger).to receive(:new) { logger }
+      invalid_project = Optimizely::Project.new('invalid', nil, spy_logger)
+      expect(invalid_project.get_feature_variable_integer('integer_single_variable_feature', 'integer_variable', user_id, user_attributes))
+        .to eq(nil)
+      expect(logger).to have_received(:log).once.with(Logger::ERROR, 'Provided datafile is in an invalid format.')
+      expect(logger).to have_received(:log).once.with(Logger::ERROR, 'Provided datafile is in an invalid format. Aborting get_feature_variable_integer.')
```
Comment (mikeng13) [inline spec/project_spec.rb]: Same as above, seems redundant

File: spec/project_spec.rb (hunk 1 above)
Comment (oakbani) [reply spec/project_spec.rb]: The first one comes from the initialize method. And the second one is logged in the relevant API method. Though duplicate, I think we should keep this at both places for clarity.
If a client is initialized, it should log the datafile error. When an API method is called, it should log exactly why the method is aborting. 
Let me know if we should revise the error log in the API method. And shorten it to only 'Aborting <method name>"

Comment (mikeng13) [reply spec/project_spec.rb]: Yeah that makes sense. I agree with your assessment.

**Pull request merged**